### PR TITLE
'Not found' middleware

### PR DESCRIPTION
**Pull request by Minstel**

### CODE REVIEW

File: src/Router/Middleware/NotFound.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    /**
+     * Class constructor
+     * 
+     * @param array $routes
+     * @param callback|int $notFound 
+     * @param callback|int $methodNotAllowed 
+     */
+    public function __construct(array $routes, $notFound = 404, $methodNotAllowed = null)
+    {
+        if (!in_array($notFound, [404, '404'], true) && !is_callable($notFound)) {
```
Comment (jasny) [inline src/Router/Middleware/NotFound.php]: You should be able to use any valid http status code (100 - 999) for `$notFound`.

File: src/Router/Middleware/NotFound.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            throw new \InvalidArgumentException("'Not found' parameter should be '404' or a callback");
+        }
+
+        if ($methodNotAllowed && !in_array($methodNotAllowed, [405, '405'], true) && !is_callable($methodNotAllowed)) {
```
Comment (jasny) [inline src/Router/Middleware/NotFound.php]: You should be able to use any valid http status code (100 - 999) for `$methodNotAllowed`.
You need to do `isset($methodNotAllowed)`, so `0` and `false` will also cause an exception.

File: src/Router/Middleware/NotFound.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+     * @param callback               $next
+     * @return ResponseInterface
+     */
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, $next = null)
+    {
+        if ($next && !is_callable($next)) {
+            throw new \InvalidArgumentException("'next' should be a callback");            
+        }
+
+        $glob = new Glob($this->routes);
```
Comment (jasny) [inline src/Router/Middleware/NotFound.php]: You mustn't create `Glob` here, `$this->routes` should already be a `Routes` object.

File: src/Router/Middleware/NotFound.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    protected $methodNotAllowed = null;
+
+    /**
+     * Class constructor
+     * 
+     * @param array $routes
+     * @param callback|int $notFound 
+     * @param callback|int $methodNotAllowed 
+     */
+    public function __construct(array $routes, $notFound = 404, $methodNotAllowed = null)
```
Comment (jasny) [inline src/Router/Middleware/NotFound.php]: `$routes` must be a `Routes` object, not an array.

File: src/Router/Middleware/NotFound.php [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                    $this->simpleResponse($response, $this->methodNotAllowed, 'Method Not Allowed') :
+                    call_user_func($this->methodNotAllowed, $request, $response);
+            }
+        } 
+
+        if (!$glob->hasRoute($request)) {
+            return is_numeric($this->notFound) ? 
+                $this->simpleResponse($response, $this->notFound, 'Not Found') :
+                call_user_func($this->notFound, $request, $response);
+        }
```
Comment (jasny) [inline src/Router/Middleware/NotFound.php:80]: No need to duplicate the code in 2 `if` blocks here:

```
if (!$this->routes->hasRoute($request)) {
   $status = $this->methodNotAllowed && $this->routes->hasRoute($request, false)
       ? $this->methodNotAllowed
       : $this->notFound;
   ...
}
```

File: src/Router/Middleware/NotFound.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+        $glob = new Glob($this->routes);          
+
+        if ($this->methodNotAllowed) {
+            $notAllowed = !$glob->hasRoute($request) && $glob->hasRoute($request, false);
+
+            if ($notAllowed) {
+                return is_numeric($this->methodNotAllowed) ? 
+                    $this->simpleResponse($response, $this->methodNotAllowed, 'Method Not Allowed') :
```
Comment (jasny) [inline src/Router/Middleware/NotFound.php]: Just always use 'Not Found' as body text, regardless of the response code.

**Review by jasny — CHANGES_REQUESTED** [review]

File: src/Router/Middleware/NotFound.php [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    /**
+     * Action for 'method not allowed' case
+     * @var string
+     **/
+    protected $methodNotAllowed = null;
+
+    /**
+     * Class constructor
+     * 
+     * @param array $routes
```
Comment (jasny) [inline src/Router/Middleware/NotFound.php]: Fix `@param Routes $routes`

File: tests/Router/Middleware/NotFoundTest.php [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+     * @dataProvider constructProvider
+     * @param string $notFound 
+     * @param string $notAllowed
+     * @param boolean $positive 
+     */
+    public function testConstruct($notFound, $notAllowed, $positive)
+    {   
+        if (!$positive) $this->expectException(\InvalidArgumentException::class);
+
+        $middleware = new NotFound(new Glob([]), $notFound, $notAllowed);
```
Comment (jasny) [inline tests/Router/Middleware/NotFoundTest.php]: Don't use `Glob`, but use a mock of `Routes`.

File: tests/Router/Middleware/NotFoundTest.php [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    /**
+     * Get requests for testing
+     *
+     * @param string $uri 
+     * @param string $method 
+     * @return array
+     */
+    public function getRequests($uri, $method)
```
Comment (jasny) [inline tests/Router/Middleware/NotFoundTest.php]: method isn't used

File: tests/Router/Middleware/NotFoundTest.php [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    {
+        $stream = $this->createMock(StreamInterface::class);
+        $stream->expects($this->once())->method('rewind');
+        $stream->expects($this->once())->method('write')->with($this->equalTo('Not Found'));
+
+        $response->method('getBody')->will($this->returnValue($stream));
+        $response->expects($this->once())->method('withBody')->with($this->equalTo($stream))->will($this->returnSelf());
+        $response->expects($this->once())->method('withStatus')->with($this->equalTo($code), $this->equalTo('Not Found'))->will($this->returnSelf());
+    }
+
```
Comment (jasny) [inline tests/Router/Middleware/NotFoundTest.php:234]: Please create tests for the case where
- the route **is** found
- the route isn't found at all
- the route isn't found, but exists for another method

**Review by jasny — CHANGES_REQUESTED** [review]
Code is good. Tests need some improvements.

File: tests/Router/Middleware/NotFoundTest.php (hunk 9 above)
Comment (Minstel) [inline tests/Router/Middleware/NotFoundTest.php]: It is.

**Comment by jasny** [comment]
@Minstel I want to merge, but... Did you push to the wrong branch? Other middelware classes have been added. Please create a branch per middleware.

**Review by jasny — APPROVED** [review]

**Pull request merged**